frontend/test/pytest/test_gradient.py

```diff
@@ -1470,6 +1470,27 @@ def my_model(data, weights, bias):
     assert np.allclose(cat_res, jax_res)
 
 
+@pytest.mark.parametrize(
+    "gate,state", ((qml.BasisState, np.array([1])), (qml.StatePrep, np.array([0, 1])))
+)
+def test_paramshift_with_gates(gate, state):
+    """Test parameter shift works with a variety of gates present in the circuit."""
+
+    dev = qml.device("lightning.qubit", wires=1)
+
+    @grad
+    @qml.qnode(dev, diff_method="parameter-shift")
+    def cost(x):
+        gate(state, wires=0)
```

Comment:
Produce garbage results I imagine (I doubt either the StatePrep or BasisState follow the two-term shift rule, which is the only one we support). The verifier should technically catch this to begin with, but as Lillian said for parameter-shift it's still very rudimentary right now.

Reply:
> The verifier should technically catch this to begin with but as Lillian said for parameter-shift it's still very rudimentary right now.

I don't like the idea that this can either produce garbage results (or possibly crash). But I understand that @josh146 wants this StatePrep and BasisState to be differentiable in this special case for the demos. :+1:

Reply:
I agree, unfortunately that's always been and still is the case for all ops that don't conform to the two-term shift rule, that's why I was pushing for the verifier so much.

Reply:
Actually the current demos don't need stateprep/basisstate to be differentiable :) it's more that the verifier shouldn't block a circuit with other differentiable gates just because stateprep/basisstate are present.

In particular:

- `BasisState` is ill-defined for differentiation (since the input is integers), and should always error if a user tries to explicitly differentiate it. I think we already do so (due to type checking that the inputs to `catalyst.grad` must be float).

- `StatePrep` _can_ be made differentiable via the [`MottonenStatePreparation`](https://docs.pennylane.ai/en/stable/code/api/pennylane.MottonenStatePreparation.html) decomposition, which decomposes the stateprep into a series of gates that act on the ground state.

Reply:
So maybe just to clarify; the bug was less so that BasisState/StatePrep were not differentiable with parameter-shift.

The bug was instead that, when switching my example code to parameter-shift, the behaviour no longer matched adjoint (I got a compiler error rather than a Verification error like I do with adjoint).

Reply:
> it's more that the verifier shouldn't block a circuit with other differentiable gates just because stateprep/basisstate are present.

True but that is a separate issue, this PR fixes the compiler error which shouldn't be there (regardless of any other issues we have) :)